File: ApexFit/Core/Configuration/ScoringConfig.swift
import Foundation

// MARK: - Root Configuration

struct ScoringConfig: Codable {
    let version: Int
    let recovery: RecoveryConfig
    let sleep: SleepConfig
    let stress: StressConfig
    let strain: StrainConfig
    let heartRateZones: HeartRateZoneConfig
    let baselines: BaselineConfig
    let healthMonitor: HealthMonitorConfig
    let sleepPlanner: SleepPlannerConfig
    let staleness: StalenessConfig
}

// MARK: - Recovery

struct RecoveryConfig: Codable {
    let weights: RecoveryWeights
    let sigmoidSteepness: Double
    let scoreRange: ScoreRange
    let zones: RecoveryZoneConfig
    let strainTargets: RecoveryStrainTargets
    let insightThresholds: RecoveryInsightThresholds
}

struct RecoveryWeights: Codable {
    let hrv: Double
    let restingHeartRate: Double
    let sleep: Double
    let respiratoryRate: Double
    let spo2: Double
    let skinTemperature: Double
}

struct RecoveryZoneConfig: Codable {
    let green: ScoreRange
    let yellow: ScoreRange
    let red: ScoreRange
}

struct RecoveryStrainTargets: Codable {
    let green: ValueRange
    let yellow: ValueRange
    let red: ValueRange
}

struct RecoveryInsightThresholds: Codable {
    let hrvPercentChange: Double
    let rhrDeltaBPM: Double
    let sleepPerformanceHigh: Double
    let sleepPerformanceLow: Double
    let skinTempDeviationCelsius: Double
}

// MARK: - Sleep

struct SleepConfig: Codable {
    let compositeWeights: SleepCompositeWeights
    let consistencyWindowNights: Int
    let consistencyDecayTau: Double
    let disturbanceScaling: Double
    let strainSupplements: [StrainSupplement]
    let debtRepaymentRate: Double
    let defaults: SleepDefaults
    let sessionDetection: SleepSessionDetection
}

struct SleepCompositeWeights: Codable {
    let sufficiency: Double
    let efficiency: Double
    let consistency: Double
    let disturbances: Double
}

struct StrainSupplement: Codable {
    let strainBelow: Double
    let addHours: Double
}

struct SleepDefaults: Codable {
    let baselineHours: Double
    let onsetLatencyMinutes: Double
}

struct SleepSessionDetection: Codable {
    let gapToleranceMinutes: Double
    let minimumDurationMinutes: Double
    let maximumNapDurationHours: Double
    let napCreditCapHours: Double
}

// MARK: - Stress

struct StressConfig: Codable {
    let weights: StressWeights
    let baselineWindowDays: Int
    let minimumBaselineSamples: Int
    let motionDampeningFactor: Double
    let bucketIntervalMinutes: Int
    let normalization: StressNormalization
    let levels: [StressLevelThreshold]
    let populationDefaults: StressPopulationDefaults
    let baevsky: BaevskyConfig
}

struct StressWeights: Codable {
    let hrv: Double
    let heartRate: Double
}

struct StressNormalization: Codable {
    let zOffset: Double
    let zRange: Double
    let outputMax: Double
}

struct StressLevelThreshold: Codable {
    let below: Double
    let label: String
}

struct StressPopulationDefaults: Codable {
    let lnRMSSDMean: Double
    let lnRMSSDStd: Double
    let heartRateMean: Double
    let heartRateStd: Double
}

struct BaevskyConfig: Codable {
    let binWidthSeconds: Double
    let normalizationDivisor: Double
    let normalizationMultiplier: Double
}

// MARK: - Strain

struct StrainConfig: Codable {
    let scalingFactor: Double
    let logOffsetConstant: Double
    let maxValue: Double
    let minValue: Double
    let zones: StrainZoneConfig
}

struct StrainZoneConfig: Codable {
    let light: ValueRange
    let moderate: ValueRange
    let high: ValueRange
    let overreaching: ValueRange
}

// MARK: - Heart Rate Zones

struct HeartRateZoneConfig: Codable {
    let boundaries: [Double]
    let multipliers: [Double]
    let sampleMaxDurationSeconds: Double
}

// MARK: - Baselines

struct BaselineConfig: Codable {
    let windowDays: Int
    let minimumSamples: Int
    let fallbackDays: Int
    let exponentialAlpha: Double
}

// MARK: - Health Monitor

struct HealthMonitorConfig: Codable {
    let zScoreThreshold: Double
}

// MARK: - Sleep Planner

struct SleepPlannerConfig: Codable {
    let goalMultipliers: SleepGoalMultipliers
}

struct SleepGoalMultipliers: Codable {
    let peak: Double
    let perform: Double
    let getBy: Double
}

// MARK: - Staleness

struct StalenessConfig: Codable {
    let heartRateHours: Int
    let sleepDataHours: Int
    let workoutHours: Int
    let restingHRHours: Int
    let hrvHours: Int
    let spo2Hours: Int
}

// MARK: - Shared Types

struct ScoreRange: Codable {
    let min: Int
    let max: Int
}

struct ValueRange: Codable {
    let min: Double
    let max: Double
}

File: android/core/model/src/main/kotlin/com/apexfit/core/model/config/ScoringConfig.kt
package com.apexfit.core.model.config

import kotlinx.serialization.Serializable

@Serializable
data class ScoringConfig(
    val version: Int,
    val recovery: RecoveryConfig,
    val sleep: SleepConfig,
    val stress: StressConfig,
    val strain: StrainConfig,
    val heartRateZones: HeartRateZoneConfig,
    val baselines: BaselineConfig,
    val healthMonitor: HealthMonitorConfig,
    val sleepPlanner: SleepPlannerConfig,
    val staleness: StalenessConfig,
)

// Recovery

@Serializable
data class RecoveryConfig(
    val weights: RecoveryWeights,
    val sigmoidSteepness: Double,
    val scoreRange: ScoreRange,
    val zones: RecoveryZoneConfig,
    val strainTargets: RecoveryStrainTargets,
    val insightThresholds: RecoveryInsightThresholds,
)

@Serializable
data class RecoveryWeights(
    val hrv: Double,
    val restingHeartRate: Double,
    val sleep: Double,
    val respiratoryRate: Double,
    val spo2: Double,
    val skinTemperature: Double,
)

@Serializable
data class RecoveryZoneConfig(
    val green: ScoreRange,
    val yellow: ScoreRange,
    val red: ScoreRange,
)

@Serializable
data class RecoveryStrainTargets(
    val green: ValueRange,
    val yellow: ValueRange,
    val red: ValueRange,
)

@Serializable
data class RecoveryInsightThresholds(
    val hrvPercentChange: Double,
    val rhrDeltaBPM: Double,
    val sleepPerformanceHigh: Double,
    val sleepPerformanceLow: Double,
    val skinTempDeviationCelsius: Double,
)

// Sleep

@Serializable
data class SleepConfig(
    val compositeWeights: SleepCompositeWeights,
    val consistencyWindowNights: Int,
    val consistencyDecayTau: Double,
    val disturbanceScaling: Double,
    val strainSupplements: List<StrainSupplement>,
    val debtRepaymentRate: Double,
    val defaults: SleepDefaults,
    val sessionDetection: SleepSessionDetection,
)

@Serializable
data class SleepCompositeWeights(
    val sufficiency: Double,
    val efficiency: Double,
    val consistency: Double,
    val disturbances: Double,
)

@Serializable
data class StrainSupplement(
    val strainBelow: Double,
    val addHours: Double,
)

@Serializable
data class SleepDefaults(
    val baselineHours: Double,
    val onsetLatencyMinutes: Double,
)

@Serializable
data class SleepSessionDetection(
    val gapToleranceMinutes: Double,
    val minimumDurationMinutes: Double,
    val maximumNapDurationHours: Double,
    val napCreditCapHours: Double,
)

// Stress

@Serializable
data class StressConfig(
    val weights: StressWeights,
    val baselineWindowDays: Int,
    val minimumBaselineSamples: Int,
    val motionDampeningFactor: Double,
    val bucketIntervalMinutes: Int,
    val normalization: StressNormalization,
    val levels: List<StressLevelThreshold>,
    val populationDefaults: StressPopulationDefaults,
    val baevsky: BaevskyConfig,
)

@Serializable
data class StressWeights(
    val hrv: Double,
    val heartRate: Double,
)

@Serializable
data class StressNormalization(
    val zOffset: Double,
    val zRange: Double,
    val outputMax: Double,
)

@Serializable
data class StressLevelThreshold(
    val below: Double,
    val label: String,
)

@Serializable
data class StressPopulationDefaults(
    val lnRMSSDMean: Double,
    val lnRMSSDStd: Double,
    val heartRateMean: Double,
    val heartRateStd: Double,
)

@Serializable
data class BaevskyConfig(
    val binWidthSeconds: Double,
    val normalizationDivisor: Double,
    val normalizationMultiplier: Double,
)

// Strain

@Serializable
data class StrainConfig(
    val scalingFactor: Double,
    val logOffsetConstant: Double,
    val maxValue: Double,
    val minValue: Double,
    val zones: StrainZoneConfig,
)

@Serializable
data class StrainZoneConfig(
    val light: ValueRange,
    val moderate: ValueRange,
    val high: ValueRange,
    val overreaching: ValueRange,
)

// Heart Rate Zones

@Serializable
data class HeartRateZoneConfig(
    val boundaries: List<Double>,
    val multipliers: List<Double>,
    val sampleMaxDurationSeconds: Double,
)

// Baselines

@Serializable
data class BaselineConfig(
    val windowDays: Int,
    val minimumSamples: Int,
    val fallbackDays: Int,
    val exponentialAlpha: Double,
)

// Health Monitor

@Serializable
data class HealthMonitorConfig(
    val zScoreThreshold: Double,
)

// Sleep Planner

@Serializable
data class SleepPlannerConfig(
    val goalMultipliers: SleepGoalMultipliers,
)

@Serializable
data class SleepGoalMultipliers(
    val peak: Double,
    val perform: Double,
    val getBy: Double,
)

// Staleness

@Serializable
data class StalenessConfig(
    val heartRateHours: Int,
    val sleepDataHours: Int,
    val workoutHours: Int,
    val restingHRHours: Int,
    val hrvHours: Int,
    val spo2Hours: Int,
)

// Shared Types

@Serializable
data class ScoreRange(
    val min: Int,
    val max: Int,
)

@Serializable
data class ValueRange(
    val min: Double,
    val max: Double,
)
